sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -2097,6 +2097,138 @@ def set_application(self, channel, appl_code):
+                    if ((1 << lane) & host_lanes_mask) == 0:
+                        continue
+                    lane = lane+1
+                    si_key_lane = "{}{}".format(si_keys, lane)
+                    val = si_settings[si_keys][si_key_lane]
+                    if ((si_keys == consts.OUTPUT_EQ_PRE_CURSOR_TARGET_RX and val <= rx_si_ctrl_advt_capability['rx_output_eq_pre_max_val']) or
+                        (si_keys == consts.OUTPUT_EQ_POST_CURSOR_TARGET_RX and val <= rx_si_ctrl_advt_capability['rx_output_eq_post_max_val']) or
+                        (si_keys == consts.OUTPUT_AMPLITUDE_TARGET_RX and val <= rx_si_ctrl_advt_capability['rx_output_amp_supported_val']) or
+                        (si_keys == consts.FIXED_INPUT_EQ_TARGET_TX and val <= tx_si_ctrl_advt_capability['tx_input_eq_max_val'])) :
+                        if (lane%2) == 0:
+                            pre_si_key_lane = "{}{}".format(si_keys, lane-1)
+                            pre_val = self.xcvr_eeprom.read(pre_si_key_lane)
+                            val = (val << 4) | pre_val
+                        self.xcvr_eeprom.write(si_key_lane, val)
+            elif (si_keys == consts.ADAPTIVE_INPUT_EQ_RECALLED_TX and 
+                 (tx_si_ctrl_advt_capability['tx_input_recall_buf1_supported'] or tx_si_ctrl_advt_capability['tx_input_recall_buf2_supported'])):
+                val = 0
+                for lane in range(self.NUM_CHANNELS):
+                    if ((1 << lane) & host_lanes_mask) == 0:
+                        continue
+                    si_key_lane = "{}{}".format(si_keys, lane+1)
+                    si_val = si_settings[si_keys][si_key_lane]
+                    lane %= (self.NUM_CHANNELS//2)
+                    mask = ~(val << (lane*2))
+                    l_data = si_val << (lane*2)
+                    val = (val & mask) | l_data
+                    self.xcvr_eeprom.write(si_key_lane, val)
+            elif ((si_keys == consts.ADAPTIVE_INPUT_EQ_ENABLE_TX and tx_si_ctrl_advt_capability['tx_input_adaptive_eq_supported']) or
+                  (si_keys == consts.CDR_ENABLE_TX and tx_si_ctrl_advt_capability['tx_cdr_supported']) or
+                  (si_keys == consts.CDR_ENABLE_RX and rx_si_ctrl_advt_capability['rx_cdr_supported'])) :
+                val = 0
+                # Read 1 byte data
+                for lane in range(self.NUM_CHANNELS):
+                    si_key_lane = "{}{}".format(si_keys, lane+1)
+                    data  = self.xcvr_eeprom.read(si_key_lane)
+                    val |= (data << lane)
+                # Write only applicable field
+                for lane in range(self.NUM_CHANNELS):
+                    if ((1 << lane) & host_lanes_mask) == 0:
+                        continue
+                    si_key_lane = "{}{}".format(si_keys, lane+1)
+                    si_val = si_settings[si_keys][si_key_lane]
+                    val &= ~(1 << lane)
+                    val |= (si_val << lane)
```

Comment:
@AnoopKamath instead of handling all SI params here in one block, can you write functions for each: CDR, pre/post cursor, output Amp?